### PR TITLE
Code for embedding extraction, t-SNE plots with hover

### PR DESCRIPTION
**Pull request by mudit1990**
@rohinikapoor @sarankrishna920 @doodhwala please review
* dal.py - Correcting the full data pull query from the database. Full data was pulled and dumped as csv. The zipped file can be found in 696DS/Data CSV Dumps folder

* model1, model3.py - added code for extracting user embeddings only for people seen in the training data beyond a particular threshold and method call for plotting t-SNE

* utils.py - added code for plotting user embeddings and email embeddings via t-SNE with hover functionality

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -9,6 +9,7 @@ __pycache__/
 # Large data files
 data/20000_emails.csv
 data/50000_emails.csv
+data/all_emails.csv
```
Comment (mohit-surana) [inline .gitignore:12]: Can you share this on Google Drive and include a link as a PR comment?

File: code/model1.py [hunk 2]
```diff
@@ -77,8 +86,33 @@ def train(self, emails, w2v):
...
                     epoch_loss += loss.data.numpy()
+            end = time.time()
+            print 'time taken ', (end-start)
             print 'loss in epoch ' + str(epoch) + ' = ' + str(epoch_loss)
 
+        print 'Number of entries in the dictionary ', len(self.emailid_train_freq)
+        email_ids, embs = self.extract_user_embeddings()
+        utils.plot_with_tsne(email_ids, embs, True)
+
+    def extract_user_embeddings(self, threshold=1):
```
Comment (mohit-surana) [inline code/model1.py]: @rohinikapoor @sarankrishna920 
If we code model2 in a way such that this works for model2 as well, 
then we can consider pulling these out into one common place instead of it being replicated across the 3 models

File: code/utils.py [hunk 3]
```diff
@@ -20,12 +25,77 @@ def get_userid(u):
...
     mapping = np.loadtxt('../resources/employee_id_mapping.csv', dtype='str', delimiter=',', skiprows=1)
     for m in mapping:
         user_id_lookup[m[0]] = int(m[1])
 
 
+def plot_with_tsne(email_ids, embeddings, display_plot):
+    """
+    expects a list of email_ids and numpy ndarray of embeddings. The numpy ndarray should have shape L,D where D is the
+    size of embeddings and L is the number of users
+    :param email2embedding:
```
Comment (mohit-surana) [inline code/utils.py]: Possibly stale docstring.

File: code/utils.py [hunk 4]
```diff
@@ -20,12 +25,77 @@ def get_userid(u):
...
+    tsne = TSNE(verbose=1, method='exact')
+    start = time.time()
+    tsne_embs = tsne.fit_transform(embeddings)
+    end = time.time()
+    print('time taken by TSNE ', (end-start))
+
+    if display_plot:
+        fig, ax = plt.subplots()
+        scatter = ax.scatter(tsne_embs[:,0], tsne_embs[:,1], s=30)
+        tooltip = mpld3.plugins.PointLabelTooltip(scatter, labels=email_ids)
```
Comment (mohit-surana) [inline code/utils.py]: A passing thought:

Let's change email_ids to labels?
And we could pass org hierarchy info along with email ids
Just need to think what all information we want to display.

File: code/utils.py [hunk 5]
```diff
@@ -20,12 +25,77 @@ def get_userid(u):
...
     """
-
+    populates the user_id_lookup table based on file employee_id_mapping.csv extracted from the db
     """
     mapping = np.loadtxt('../resources/employee_id_mapping.csv', dtype='str', delimiter=',', skiprows=1)
     for m in mapping:
         user_id_lookup[m[0]] = int(m[1])
 
 
+def plot_with_tsne(email_ids, embeddings, display_plot):
```
Comment (mohit-surana) [inline code/utils.py]: Do we need a display_plot flag here?
We could set it to have a default True value

**Review by mohit-surana — APPROVED** [review]
This looks good and we should merge so that everyone can use t-sne and run experiments.
Minor comments have been added that can be addressed later on (not a priority)
:shipit:

File: .gitignore (hunk 1 above)
Comment (mudit1990) [reply .gitignore:12]: done

File: code/utils.py (hunk 3 above)
Comment (mudit1990) [reply code/utils.py]: done

File: code/utils.py (hunk 4 above)
Comment (mudit1990) [reply code/utils.py]: done

File: code/utils.py (hunk 5 above)
Comment (mudit1990) [reply code/utils.py]: we need it because mpld3 cannot display multiple plots, but someone may want to run just tSNE after every epoch. Yes, I will put the default value as true
